File: array/rotate_array.py
# Python version of the Java solution for Rotate array
class Solution(object):
    def rotate(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: None Do not return anything, modify nums in-place instead.
        """
        # for loop through array
        # index + 3

        #if index + k is more than size of array - 1
        # k - (size of array - index ) 
        if len(nums) == 1 or k == 0:
            nums = nums
        #swap 
        else:
            k %= len(nums)
            if k == 0:
                return nums
            else:
              # last k elements [-k:] becomes first k elements [:k]
              # the remaining elements are appended
                nums[:k], nums[k:] = nums[-k:], nums[:-k]

# Alternative solution: more concise

class Solution(object):
    def rotate(self, nums, k):
      n = len(nums)
      k = k%n
      nums[:] = nums[n-k:] + nums[:n-k]
                

File: array/rotate_array.java
// Name: Rotate Array
// Problem: Given an array and value k, rotate the array to the right k times

class Solution {
    public void rotate(int[] nums, int k) {
        k = k % nums.length;  // Normalize k

        if (k != 0) {
            int n = nums.length;
            // copy last k elements
            int[] copyOfEnd = Arrays.copyOfRange(nums, nums.length-k, nums.length); 
            // copy first element index 0 to k
            int[] copyOfStart = Arrays.copyOfRange(nums, 0, nums.length-k);
            // merge arrays from 0 to k for first array and k to length of array - k
            System.arraycopy(copyOfEnd, 0, nums, 0, k);
            System.arraycopy(copyOfStart, 0, nums, k, n-k);
    }
    }
}
